controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,887 @@
+
+  // Perform a power-up and calibration sequence on
+  // the A/D converter
+  ADC_Regs *adc = ADC_BASE;
+
+  // Exit deep power down mode and turn on the
+  // internal voltage regulator.
+  adc->adc[0].ctrl = 0x10000000;
+
+  // Wait for the startup time specified in the datasheet
+  // for the voltage regulator to become ready.
+  // The time in the datasheet is 20 microseconds, but
+  // I'll wait for 30 just to be extra conservative
+  BusyWait(30);
+
+  // Calibrate the A/D for single ended channels
+  adc->adc[0].ctrl |= 0x80000000;
+
+  // Wait until the CAL bit is cleared meaning
+  // calibration is finished
+  while (adc->adc[0].ctrl & 0x80000000) {
+  }
+
+  // Clear all the status bits
+  adc->adc[0].stat = 0x3FF;
+
+  // Enable the A/D
+  adc->adc[0].ctrl |= 0x00000001;
+
+  // Wait for the ADRDY status bit to be set
+  while (!(adc->adc[0].stat & 0x00000001)) {
+  }
+
+  // Configure the A/D as 12-bit resolution
+  adc->adc[0].cfg[0] = 0x00000000;
+
+  // Set sample time. I'm using 92.5 A/D clocks (a little over 1us)
+  // to sample.  We'll need to do a bit of testing to see what the
+  // best value is for this
+  adc->adc[0].samp[0] = 5;
+}
+
+// Read a single A/D channel.
+// This just does the conversion immediately when called.
+//
+// It would be better to start a sequence of conversions at the beginning of the
+// loop but the HAL interface currently isn't set up to run that way.  That's an
+// improvement that can be made later
+int HalApi::analogRead(AnalogPin pin) {
+
+  int channel = 0;
+  switch (pin) {
+  case AnalogPin::PATIENT_PRESSURE:
+    channel = 6;
+    break;
+
+  case AnalogPin::INFLOW_PRESSURE_DIFF:
+    channel = 9;
+    break;
+
+  case AnalogPin::OUTFLOW_PRESSURE_DIFF:
+    channel = 15;
+    break;
+  }
+
+  ADC_Regs *adc = ADC_BASE;
+  adc->adc[0].seq[0] = channel << 6;
+
+  // Clear the EOC flag
+  adc->adc[0].stat = 4;
+
+  // Start the conversion
+  adc->adc[0].ctrl |= 4;
+
+  // Wait for the end of the conversion
+  while (!(adc->adc[0].stat & 4)) {
+  }
+
+  // Return the result
+  return adc->adc[0].data;
+}
+
+/******************************************************************
+ * PWM outputs
+ *
+ * The following four outputs could be driven
+ * as PWM outputs:
+ *
+ * PA8  - Timer 1 Channel 1 - heater control
+ * PA11 - Timer 1 Channel 4 - solenoid
+ * PB3  - Timer 2 Channel 2 - blower control
+ * PB4  - Timer 3 Channel 1 - buzzer
+ *
+ * For now I'll just set up the blower since that's the only
+ * one called out in the HAL
+ *
+ * These timers are documented in chapters 26 and 27 of the reference
+ * manual.
+ *****************************************************************/
+static void InitPwmOut() {
+  // The PWM frequency isn't mentioned anywhere that I can find, so
+  // I'm just picking a reasonable number.  This can be refined later
+  const int pwmFreqHz = 20000;
+
+  EnableClock(TIMER2_BASE);
+
+  // Connect PB3 to timer 2
+  GPIO_PinAltFunc(GPIO_B_BASE, 3, 1);
```

Comment:
No, it's really an index.  The alternate function is a setting and the various functions are are different for each pin.  I don't think an enum really works well here, it would have meaningless names like ALT_0, ALT_1, ALT_2 ...  That seems kind of silly.